nvflare/apis/utils/reliable_message.py

```diff
@@ -295,15 +295,35 @@ def _receive_request(cls, topic: str, request: Shareable, fl_ctx: FLContext):
                 # no handler registered for this topic!
                 cls.error(fl_ctx, f"no handler registered for request {rm_topic=}")
                 return make_reply(ReturnCode.TOPIC_UNKNOWN)
-            if cls._req_completed.get(tx_id):
-                cls.debug(fl_ctx, "Completed tx_id received")
-                return _status_reply(STATUS_DUP_REQUEST)
-            receiver = cls._get_or_create_receiver(rm_topic, request, handler_f)
+
+            # check whether the request is still standing or completed
+            # we should check to get the receiver first, and check req_completed next:
+            # if the receiver does not exist in _req_receivers and already completed,
+            # then it must exist in _req_completed (since we put it in _req_completed before removing it
+            # from _req_receivers).
+            receiver = cls._req_receivers.get(tx_id)
+            if not receiver:
+                # no standing process for this request
+                # further check whether this request was already completed
+                if cls._req_completed.get(tx_id):
+                    # this request was already completed!
+                    cls.debug(fl_ctx, "Completed tx_id received")
+                    return _status_reply(STATUS_DUP_REQUEST)
+
+            if not receiver:
+                # this is a valid new request
+                receiver = cls._get_or_create_receiver(rm_topic, request, handler_f)
```

Comment:
this part can be put inside previous block since they are under same condition "if not receiver"

Reply:
I kept them separate because they are for different purposes.